Add Altair validator guide and initial p2p spec (#2201)
ethereum/eth2.0-specs: specs/altair/validator.md

A validator should plan for future sync committee assignments by noting which sync committee periods they are selected for participation.
Specifically, a validator should:
* Upon (re)syncing the chain and upon sync committee period boundaries, check for assignments in the current and next sync committee periods.
* If the validator is in the current sync committee period, they can perform the responsibilities below for sync committee rewards.
* If the validator is in the next sync committee period, they should wait until the next `EPOCHS_PER_SYNC_COMMITTEE_PERIOD` boundary and then perform the responsibilities throughout that period.

## Beacon chain responsibilities

A validator maintains the responsibilities given in the Phase 0 document.

Block proposals are modified to incorporate the sync committee signatures as detailed below.

When assigned to a sync committee, validators have a new responsibility to sign beacon block roots during each slot of the sync committee period.
These signatures are aggregated and routed to the proposer over gossip for inclusion into a beacon block. 

> Assignments to a particular sync committee are infrequent at normal validator counts; however, an action every slot is be required when in the currently active sync committee.

Reviewer: ```suggestion
Assignments to a particular sync committee are infrequent at normal validator counts; however, an action every slot is be required when in the current active sync committee.
```

#### Constructing the `BeaconBlockBody`

Each section of [Constructing the `BeaconBlockBody`](../phase0/validator.md#constructing-the-beaconblockbody) should be followed. 
After constructing the `BeaconBlockBody` as per that section, the proposer has an additional task to include the sync committee signatures:

##### Sync committee

The proposer receives a number of `SyncCommitteeContribution`s (wrapped in `SignedContributionAndProof`s on the wire) from validators in the sync committee who are selected to partially aggregate signatures from the full committee (see below for details).

The proposer collects these contributions for further aggregation when preparing a block. 
Recall `block.body.sync_aggregate.sync_committee_bits` is a `Bitvector` where the `i`th bit is `True` if the corresponding validator in the sync committee has produced a valid signature, and that `block.body.sync_aggregate.sync_committee_signature` is the aggregate BLS signature combining all of the valid signatures.

Given a collection of `contributions` and the `BeaconBlock` under construction, the proposer processes them as follows:

```python

> def process_sync_committee_contributions(state: BeaconState, 

Reviewer: If they use this function, you need to dictate that contributions *must* be non-overlapping

Reviewer: Might be easiest to dictate that they take the *best* contribution from each subcommittee

Author: ah yes, when i wrote this i was thinking we would end up w/ contiguous subcommittees which we don't have with the current definition of `compute_subnet_for_sync_committee`.  i think this functions needs to be updated; if we use the aggregates in the `state` we will move to non-overlapping subsets by default so i think best to leave this function alone

                                 privkey: int) -> SyncCommitteeSignature:
    epoch = get_current_epoch(state)
    domain = get_domain(state, DOMAIN_SYNC_COMMITTEE, epoch)
    signing_root = compute_signing_root(get_block_root_at_slot(state, state.slot), domain)
    signature = bls.Sign(privkey, signing_root)

    return SyncCommitteeSignature(slot=state.slot, validator_index=validator_index, signature=signature)
```

##### Broadcast sync committee signature

The validator broadcasts the assembled signature to the assigned subnet, the `sync_committee_{subnet_id}` pubsub topic.

The `subnet_id` is derived from the position in the sync committee such that the sync committee is divided into "subcommittees".
It can be computed via `compute_subnet_for_sync_committee` where `state` is a `BeaconState` during the matching sync committee period:

```python
def compute_subnet_for_sync_committee(state: BeaconState, validator_index: ValidatorIndex) -> uint64:
    pubkey = state.validators[validator_index].pubkey
    sync_committee_index = state.current_sync_committee.pubkeys.index(pubkey)
    return uint64(sync_committee_index % SYNC_COMMITTEE_SUBNET_COUNT)
```

> *NOTE*: subnet assignment does not change during the duration of a validator's assignment to a given sync committee.

Reviewer: Capitalize the first word after "Note". Also "Note" should not be all caps

```suggestion
*Note*: Subnet assignment does not change during the duration of a validator's assignment to a given sync committee.
```

> *NOTE*: a validator joins the "first" subnet (in sync committee index order) if they are duplicated in a given committee; this fact is reflected in the semantics of Python's `index` function in the definition for `compute_subnet_for_sync_committee`.

Author: resolved in updated scheme

Given all of the (valid) collected `sync_committee_signatures` from the `sync_committee_{subnet_id}` gossip during the selected `slot` with an equivalent `beacon_block_root` to that of the aggregator, the aggregator creates a `contribution: SyncCommitteeContribution` with the following fields:

###### Slot

Set `contribution.slot = state.slot` where `state` is the `BeaconState` for the slot in question.

###### Aggregate bits

Let `contribution.aggregate_bits` be a `Bitvector[SYNC_COMMITTEE_SIZE]`, where the `index`th bit is set in the `Bitvector` for each corresponding validator included in this aggregate.
Note that an aggregator needs to find the index in the sync committee (as returned by `get_sync_committee_indices`) for a given validator referenced by `sync_committee_signature.validator_index` and _this_ is the index to set in the `Bitvector`. 
For example, validator with index `2044` could be at index `3` in the current sync committee so position `3` in the `Bitvector` is set if a signature from this validator is included in the aggregation. 

> Also note that a validator could be included multiple times in a given sync committee such that multiple bits are set for a single `SyncCommitteeSignature`.

Reviewer: Ah, I see what you are doing here... The aggregator can include for positions outside of the sub committee.

I *think* that subcommittee/aggregate validations will be simpler if we restrict and have validators double broadcast to different subnets.

In the previous aggregation scheme (no subcommittee aggregates), double broadcasting didn't make much sense to me but in the context of multiple sub committees, I think it does.

Note, if you are duplicated within a single subcommittee, aggregator should include both

Author: resolved in newer scheme

###### Signature

Set `contribution.signature = aggregate_signature` where `aggregate_signature` is obtained from:

```python
def get_aggregate_signature(sync_committee_signatures: Set[SyncCommitteeSignature]) -> BLSSignature:
    signatures = [sync_committee_signature.signature for sync_committee_signature in sync_committee_signatures]
    return bls.Aggregate(signatures)
```

> Note that if a validator is included multiple times in a given sync committee the aggregator must include their signature multiple times in the argument to `get_aggregate_signature` as well.

Reviewer: Use "Note:" notation

    aggregation_bits: Bitvector[SYNC_COMMITTEE_SIZE // SYNC_COMMITTEE_SUBNET_COUNT]
    # Signature by the validator(s) over the block root of `slot`
    signature: BLSSignature
```

### `ContributionAndProof`

```python
class ContributionAndProof(Container):
    aggregator_index: ValidatorIndex
    contribution: SyncCommitteeContribution
    selection_proof: BLSSignature
```

### `SignedContributionAndProof`

```python
class SignedContributionAndProof(Container):
    message: ContributionAndProof
    signature: BLSSignature
```

## Validator assignments

A validator determines beacon committee assignments and beacon block proposal duties as defined in the Phase 0 document.

### Sync Committee

To determine sync committee assignments, a validator can run the following function: `is_assigned_to_sync_committee(state, epoch, validator_index)` where `epoch` is an epoch number within the current or next sync committee period.
This function is a predicate indicating the presence or absence of the validator in the corresponding sync committee for the queried sync committee period.

```python
def compute_sync_committee_period(epoch: Epoch) -> uint64:

>     return epoch // EPOCHS_PER_SYNC_COMMITTEE_PERIOD

Reviewer 2: What if the fork epoch does not align with the sync committee period? Will we have a shorter period at the start?

Author: that seems to be the case, don't think it is worth extra complexity to adjust in protocol. if we have security concerns here we can add <24hrs of epochs to the fork epoch and just wait until it is aligned

After constructing the `BeaconBlockBody` as per that section, the proposer has an additional task to include the sync committee signatures:

##### Sync committee

The proposer receives a number of `SyncCommitteeContribution`s (wrapped in `SignedContributionAndProof`s on the wire) from validators in the sync committee who are selected to partially aggregate signatures from independent subcommittees formed by breaking the full sync committee into `SYNC_COMMITTEE_SUBNET_COUNT` pieces (see below for details).

The proposer collects these contributions for further aggregation when preparing a block. 
Proposers should select the best contribution seen across all aggregators for each subnet/subcommittee when preparing a block. 
A contribution with more valid signatures is better than a contribution with fewer signatures.

Recall `block.body.sync_aggregate.sync_committee_bits` is a `Bitvector` where the `i`th bit is `True` if the corresponding validator in the sync committee has produced a valid signature, and that `block.body.sync_aggregate.sync_committee_signature` is the aggregate BLS signature combining all of the valid signatures.

> Given a collection of the best seen `contributions` (with no repeating `subcommittee_index` values) and the `BeaconBlock` under construction, the proposer processes them as follows:

Reviewer 2: > with no repeating `subcommittee_index` values

This is not checked in the function below. Is it necessary? The below review comment was resolved, but it doesn't look like it was addressed, or is this different?

Author: we could allow repeating `subcommittee_index` values as long as the `aggregation_bits` are non-overlapping

the validation is not checked in the function as this prose is supposed to stand-in for a precondition to the function

###### Subcommittee index

Set `contribution.subcommittee_index` to the index for the subcommittee index corresponding to the subcommittee assigned to this subnet. This index matches the `subnet_id` used to derive the topic name.

###### Aggregate bits

Let `contribution.aggregation_bits` be a `Bitvector[SYNC_COMMITTEE_SIZE // SYNC_COMMITTEE_SUBNET_COUNT]`, where the `index`th bit is set in the `Bitvector` for each corresponding validator included in this aggregate from the corresponding subcommittee.
An aggregator needs to find the index in the sync committee (as returned by `get_sync_committee_indices`) for a given validator referenced by `sync_committee_signature.validator_index` and map the sync committee index to an index in the subcommittee (along with the prior `subcommittee_index`). This index within the subcommittee is the one set in the `Bitvector`.
For example, a validator with index `2044` could be at index `15` in the current sync committee. This sync committee index maps to `subcommittee_index` `1` with position `7` in the `Bitvector` for the contribution. 

> Also note that a validator could be included multiple times in a given subcommittee such that multiple bits are set for a single `SyncCommitteeSignature`.

Reviewer 2: ```suggestion
Also note that a validator **could be included multiple times** in a given subcommittee such that multiple bits are set for a single `SyncCommitteeSignature`.
```
Nitpick: This is different than any existing committee logic, worth emphasizing in my opinion.

This previous document is referred to below as the "Phase 0 document".

Altair introduces a new type of committee: the sync committee. Sync committees are responsible for signing each block of the canonical chain and there exists an efficient algorithm for light clients to sync the chain using the output of the sync committees. 
See the [sync protocol](./sync-protocol.md) for further details on the light client sync. 
Under this network upgrade, validators track their participation in this new committee type and produce the relevant signatures as required. 
Block proposers incorporate the (aggregated) sync committee signatures into each block they produce.

## Prerequisites

All terminology, constants, functions, and protocol mechanics defined in the [Altair -- The Beacon Chain](./beacon-chain.md) doc are requisite for this document and used throughout. 
Please see this document before continuing and use as a reference throughout.

## Warning

This document is currently illustrative for early Altair testnets and some parts are subject to change. 

> In particular, the sync committee aggregation scheme may change to reduce resource consumption and better leverage the subcommittee aggregates defined in the state. 

Author: i'll keep for now (e.g. may think more about gossip validations as we go, i went quickly through those w/ new changes) but will make less specific

##### Broadcast sync committee contribution

If the validator is selected to aggregate (`is_sync_committee_aggregator`), then they broadcast their best aggregate as a `SignedContributionAndProof` to the global aggregate channel (`sync_committee_contribution_and_proof` topic) two-thirds of the way through the `slot`-that is, `SECONDS_PER_SLOT * 2 / 3` seconds after the start of `slot`.

Selection proofs are provided in `ContributionAndProof` to prove to the gossip channel that the validator has been selected as an aggregator.

`ContributionAndProof` messages are signed by the aggregator and broadcast inside of `SignedContributionAndProof` objects to prevent a class of DoS attacks and message forgeries.

First, `contribution_and_proof = get_contribution_and_proof(state, validator_index, contribution, privkey)` is constructed.

```python
def get_contribution_and_proof(state: BeaconState,
                               aggregator_index: ValidatorIndex,
                               contribution: SyncCommitteeContribution,
                               privkey: int) -> ContributionAndProof:

>     selection_proof = get_sync_committee_slot_signature(state, 

Author: yep, wasnt sure if we did that or followed styling for function params; ill fix

Subnet assignments are known `EPOCHS_PER_SYNC_COMMITTEE_PERIOD` epochs in advance and can be computed with `compute_subnets_for_sync_committee` defined above. 
ENR advertisement is indicated by setting the appropriate bit(s) of the bitfield found under the `syncnets` key in the ENR corresponding to the derived `subnet_id`(s). 
Any bits modified for the sync committee responsibilities are unset in the ENR after any validators have left the sync committee.

*Note*: The first sync committee from phase 0 to the Altair fork will not be known until the fork happens which implies subnet assignments are not known until then. 
Early sync committee members should listen for topic subscriptions from peers and employ discovery via the ENR advertisements near the fork boundary to form initial subnets 
Some early sync committee rewards may be missed while the initial subnets form.

* To join a sync committee subnet, select a random number of epochs before the end of the current sync committee period between 1 and `SYNC_COMMITTEE_SUBNET_COUNT`, inclusive. 
Validators should join their member subnet at the beginning of the epoch they have randomly selected. 

> For example, if the next sync committee period starts at epoch `853,248` and the validator randomly selects an offset of `3`, they should join the subnet at the beginning of epoch `853,245`. 

Reviewer 2: Should this be 853,251? I.e. add 3 instead of subtract 3?

    slot: Slot
    # Block root for this contribution
    beacon_block_root: Root
    # The subcommittee this contribution pertains to out of the broader sync committee
    subcommittee_index: uint64
    # A bit is set if a signature from the validator at the corresponding
    # index in the subcommittee is present in the aggregate `signature`.

>     aggregate_bits: Bitvector[SYNC_COMMITTEE_SIZE // SYNC_COMMITTEE_SUBNET_COUNT]

Reviewer: prefer `aggregation_bits` to match naming in `Attestation`

##### Sync committee

> The proposer receives a number of `SyncCommitteeContribution`s (wrapped in `SignedContributionAndProof`s on the wire) from validators in the sync committee who are selected to partially aggregate signatures from individual subcommittees formed by breaking the full sync committee into `SYNC_COMMITTEE_SUBNET_COUNT` pieces (see below for details).

Reviewer: "independent" might be a bit better descriptive to illustrate that they don't overlap

```suggestion
The proposer receives a number of `SyncCommitteeContribution`s (wrapped in `SignedContributionAndProof`s on the wire) from validators in the sync committee who are selected to partially aggregate signatures from independent subcommittees formed by breaking the full sync committee into `SYNC_COMMITTEE_SUBNET_COUNT` pieces (see below for details).
```

The proposer collects these contributions for further aggregation when preparing a block. 

> Proposers should select the best contribution seen across all aggregators for a given subnet/subcommittee when preparing a block. 

Reviewer: ```suggestion
Proposers should select the best contribution seen across all aggregators for each subnet/subcommittee when preparing a block. 
```

A contribution with more valid signatures is better than a contribution with fewer signatures.

Recall `block.body.sync_aggregate.sync_committee_bits` is a `Bitvector` where the `i`th bit is `True` if the corresponding validator in the sync committee has produced a valid signature, and that `block.body.sync_aggregate.sync_committee_signature` is the aggregate BLS signature combining all of the valid signatures.

Given a collection of the best seen `contributions` (with no repeating `subcommittee_index` values) and the `BeaconBlock` under construction, the proposer processes them as follows:

```python
def process_sync_committee_contributions(state: BeaconState, 
                                         block: BeaconBlock, 
                                         contributions: Set[SyncCommitteeContribution]) -> None:
    sync_aggregate = SyncAggregate()
    signatures = []

    for contribution in contributions:
        subcommittee_index = contribution.subcommittee_index
        for index, participated in enumerate(contribution.aggregate_bits):
            if participated:

>                 participant_index = SYNC_COMMITTEE_SUBNET_COUNT * subcommittee_index + index

Reviewer: I think `SYNC_COMMITTEE_SUBNET_COUNT` is incorrect. Shouldn't this multiple by `size_of_subcommittees` rather than number of subcommittees

Author: nice catch

The validator broadcasts the assembled signature to the assigned subnet, the `sync_committee_{subnet_id}` pubsub topic.

The `subnet_id` is derived from the position in the sync committee such that the sync committee is divided into "subcommittees".
It can be computed via `compute_subnets_for_sync_committee` where `state` is a `BeaconState` during the matching sync committee period. This function returns multiple subnets if a given validator index is included multiple times in a given sync committee across multiple subcommittees.

```python
def compute_subnets_for_sync_committee(state: BeaconState, validator_index: ValidatorIndex) -> Sequence[uint64]:
    target_pubkey = state.validators[validator_index].pubkey
    sync_committee_indices = [index for index, pubkey in enumerate(state.current_sync_committee.pubkeys) if pubkey == target_pubkey]
    return [

>       uint64(index // (SYNC_COMMITEE_SIZE // SYNC_COMMITTEE_SUBNET_COUNT))

Reviewer: indent level

If a validator is selected to aggregate the `SyncCommitteeSignature`s produced on a subnet during a given `slot`, they construct an aggregated `SyncCommitteeContribution`. 

Given all of the (valid) collected `sync_committee_signatures: Set[SyncCommitteeSignature]` from the `sync_committee_{subnet_id}` gossip during the selected `slot` with an equivalent `beacon_block_root` to that of the aggregator, the aggregator creates a `contribution: SyncCommitteeContribution` with the following fields:

###### Slot

Set `contribution.slot = state.slot` where `state` is the `BeaconState` for the slot in question.

###### Beacon block root

Set `contribution.beacon_block_root = beacon_block_root` from the `beacon_block_root` found in the `sync_committee_signatures`.

###### Subcommittee index

> Set `contribution.subcommittee_index` to the index for the subcommittee index corresponding to the subcommittee assigned to this subnet. This index will match the `subnet_id` used to derive the topic name.

Reviewer: ```suggestion
Set `contribution.subcommittee_index` to the index for the subcommittee index corresponding to the subcommittee assigned to this subnet. This index matches the `subnet_id` used to derive the topic name.
```

*Note*: The first sync committee from phase 0 to the Altair fork will not be known until the fork happens which implies subnet assignments are not known until then. Early sync committee members should listen for topic subscriptions from peers and employ discovery via the ENR advertisements near the fork boundary to form initial subnets. Some early sync committee rewards may be missed while the initial subnets form.

> * To join a sync committee subnet, select a random number of epochs before the end of the current sync committee period between 1 and `SYNC_COMMITTEE_SUBNET_COUNT`, inclusive. Validators should join their member subnet at the beginning of the epoch they have randomly selected. For example, if the next sync committee period starts at epoch `853,248` and the validator randomly selects an offset of `3`, they should join the subnet at the beginning of epoch `853,245`. Validators should leverage the lookahead period on sync committee assignments so that they can join the appropriate subnets ahead of their assigned sync committee period.

Reviewer 2: Nitpick: can we keep the lines short to reduce diff problems?

Author: yeah i was applying the "only change if i touch it" process but i can go thru this PR before it is final and change them

Author: ah actually, here i think it will break the `*` bullet formatting to have a newline -- do you know?
